packages/gatsby/src/utils/page-data.js

```diff
@@ -22,7 +25,71 @@ const write = async ({ publicDir }, page, result) => {
   await fs.outputFile(filePath, JSON.stringify(body))
 }
 
+/**
+ * Returns the corresponding html file path for a page-data.json file
+ */
+const getPageDataHtmlPath = pageDataPath =>
+  pageDataPath
+    .replace(`/index/`, `/`)
+    .replace(`./public/page-data/`, `/`)
+    .replace(`/page-data.json`, ``)
+
+/**
+ * Finds a list of unused page-data.json files that are leftover from deleted pages
+ *
+ * @argument {object}
+ * {object} availablePagePaths: an array of paths to pages
+ * {object} directory: the root of the Gatsby site
+ * @returns {array}
+ */
+const getUnusedPageDataPaths = ({ availablePagePaths, directory }) => {
+  const allPageDataFiles = glob.sync(`./public/page-data/**/page-data.json`, {
+    cwd: directory,
+  })
+
+  const unusedPageDataFiles = allPageDataFiles.filter(pageDataPath => {
+    // string replaces allow us to match page-data paths against public html paths
+    const normalizedPageDataPath = getPageDataHtmlPath(pageDataPath)
+
+    // if this page-data.json file doesn't have a matching html file,
+    // it's not being used
+    const thisPageDataIsUnused =
+      !availablePagePaths.includes(normalizedPageDataPath) &&
```

Comment:
This will be super slow when you have a large page count. Can this be a Set?

Reply:
This also is occurring for /every/ page twice, so it's searching every page twice, for every page. A Set is definitely a very good idea here :P

Reply:
@pieh and I manually tested this with a 25000 page test site and found that when timing calculating which pages needed to be deleted vs timing actually deleting them, it actually didn't take much time to calculate but took much more time deleting. I can make that change though. I've never used Set's before so seems like a good opportunity to learn about them.

Reply:
I've updated this to use Set.has() instead